content/browser/attribution_reporting/attribution_storage_sql.cc

```diff
@@ -2999,6 +3388,272 @@ bool AttributionStorageSql::StoreAttributionReport(AttributionReport& report) {
+
+  // Check if all remaining budgets are enough for required_budget
+  for (const auto& pair : remaining_budgets) {
+    if (pair.second < required_budget) {
+      return AggregatableResult::kInsufficientBudget;
+    }
+  }
+
+  // Insert epochs not already stored
+  static constexpr char kInsertEpochOrigin[] =
+    "INSERT INTO per_origin_filters"
+    "(epoch,origin,initial_budget,consumed_budget)"
+    "VALUES(?,?,?,?)";
+  
+  for (auto& epoch : epochs_to_insert) {
+    sql::Statement stmt(db_.GetCachedStatement(SQL_FROM_HERE, kInsertEpochOrigin));
+    stmt.BindInt64(0, epoch);
+    stmt.BindString(1, net::SchemefulSite(querying_origin).Serialize());
+    stmt.BindDouble(2, kInitialBudget);
+    stmt.BindDouble(3, required_budget);
+    if (!stmt.Run()) {
+      return AggregatableResult::kInternalError;
+    }  
+  }
+  
+  // Update epochs already stored
+  static constexpr char kAdjustBudgetConsumed[] =
+    "UPDATE per_origin_filters "
+    "SET consumed_budget=consumed_budget+?"
+    "WHERE epoch=? and origin=?";
+
+  for (auto& epoch : epochs_to_update) {
+    sql::Statement stmt(db_.GetCachedStatement(SQL_FROM_HERE, kAdjustBudgetConsumed));
+    stmt.BindDouble(0, required_budget);
+    stmt.BindInt64(1, epoch);
+    stmt.BindString(2, net::SchemefulSite(querying_origin).Serialize());
+    if (!stmt.Run()) {
+      return AggregatableResult::kInternalError;
+    }
+  }
+
+  // TODO(kelly): optimize, don't do insertions/updates in a loop?
+  
+  // Finishing atomic operation
+  if (!transaction.Commit()) {
+    return AggregatableResult::kInternalError;
+  }
+
+  return AggregatableResult::kSuccess;
+}
+
+
+AggregatableResult
+AttributionStorageSql::MaybeStoreAggregatableAttributionReportDataM2M(
+    const AttributionInfo& attribution_info,
+    std::vector<Partition>& partitions,
+    std::optional<AttributionReport>& report,
+    const AttributionTrigger& trigger,
+    StoredSource& source /*for retrocompatibility*/) {
+
+  // Computes the budget-required per requested epoch for the querying origin using 
+  // different optimization methods. Tries to consume budget from every requested epoch from the
+  // querying origin. Partitions whose epochs don't have remaining budget will have their reports NULLed.
+  // Sums the reports across partitions to create one attribution report.
+
+  const SuitableOrigin& querying_origin = trigger.destination_origin();
+
+  const attribution_reporting::TriggerRegistration& trigger_registration =
+      trigger.registration();
+
+  const attribution_reporting::AggregatableValues& aggregatable_cap_values = 
+      trigger_registration.aggregatable_cap_values;
+
+  const double global_epsilon = trigger_registration.global_epsilon;
+
+  // Compute global sensitivity
+  double global_sensitivity = 0;
+  if (std::strcmp(kSensitivityMetric, "L1") == 0) {
+    for (auto& pair : aggregatable_cap_values.values()) {
+      global_sensitivity += pair.second;
+    }
```

Comment:
Is this what the original ARA code uses to compute their attribution budget? They don't use epsilon and directly track a global sensitivity budget based on the `aggregatable_cap_values`, right?

Reply:
No there is no such thing as "aggregatable_cap_values". I added this argument. 
They use a cap/max-contribution value that is fixed in ARA which they divide across all mini reports. 
They call this "aggregatable_consumed_budget" and they deduct that from the budget of the source that was matched to the report. e.g. the source's budget capacity could be 4000 then for the mini reports they deduct the fixed cap e.g. 2000

Reply:
Nice, thanks! Indeed, they do event-level DP so they can deduct from the source I guess.